### DISCUSSION TITLE
Adding support for DDS-238-2-wifi

### DISCUSSION THREAD

**Discussion by davoau** [arendst/Tasmota#15658] [Ideas]
Hi,
I am wanting to add support for a the DDS-238-2-wifi energy monitor, specifically to read the reverse energy data.
Currently we use the following config to read tuyaMCU.
```
{"NAME":"Hiking DDS238-2 WIFI","GPIO":[0,2272,0,2304,0,0,0,0,0,0,320,0,32,0],"FLAG":0,"BASE":54}
```
Info here:
https://github.com/krikk/Hiking-DDS238-2-WIFI-Din-Rail-Energy-Meter-flashing-Tasmota

But it doesn't show frequency or reverse power - which I would like to show power being sent to grid from my solar system.

There is a DDS-238-2 modbus version which is supported, and has it own driver: https://github.com/arendst/Tasmota/blob/master/tasmota/xnrg_09_dds2382.ino.

It has the following code to read the registers:
```
      //           0     1     2     3     4     5     6     7     8     9     A     B     C     D     E     F    10    11           = ModBus register
      //  0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40  = Buffer index
      // SA FC BC EnergyTotal             ExportActiv ImportActiv                         Volta Curre APowe RPowe PFact Frequ Crc--  = DDS238-2 ZN/S version 1 (#6384)
      // SA FC BC EnergyTotal                                     ExportActiv ImportActiv Volta Curre APowe RPowe PFact Frequ Crc--  = DDS238-2 ZN/S version 2 (#6531)

      Energy.voltage[0] = (float)((buffer[27] << 8) + buffer[28]) / 10.0;
      Energy.current[0] = (float)((buffer[29] << 8) + buffer[30]) / 100.0;
      Energy.active_power[0] = (float)((buffer[31] << 8) + buffer[32]);
      Energy.reactive_power[0] = (float)((buffer[33] << 8) + buffer[34]);
      Energy.power_factor[0] = (float)((buffer[35] << 8) + buffer[36]) / 1000.0;                                          // 1.00
      Energy.frequency[0] = (float)((buffer[37] << 8) + buffer[38]) / 100.0;                                              // 50.0 Hz
      uint8_t offset = 11;
      if (Settings.flag3.dds2382_model) {
        offset = 19;
      }
      Energy.export_active = (float)((buffer[offset] << 24) + (buffer[offset +1] << 16) + (buffer[offset +2] << 8) + buffer[offset +3]) / 100.0;  // 429496729.0 W
      float import_active = (float)((buffer[offset +4] << 24) + (buffer[offset +5] << 16) + (buffer[offset +6] << 8) + buffer[offset +7]) / 100.0;  // 429496729.0 W
```

The below register info is from the factory that makes the DDS-238-2 wifi and modbus versions:
```
12 registers
24 bytes

  //           0     1     2     3     4     5     6     7     8     9     A     B     C     D     E     F    10    11           = ModBus register
  //  0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40  = Buffer index
  // 01 03 24 00 00 1C 7B 00 00 00 00 00 00 00 00 00 00 00 00 00 00 04 F3 00 00 17 88 09 77 01 A6 03 F8 00 70 03 E1 13 8A 63 ED
  // 01 03 24 00 00 1C AD 00 00 00 00 00 00 00 00 00 00 00 00 00 00 04 F3 00 00 17 BA 09 77 01 F1 04 AF 00 6C 03 E3 13 86 41 A2
  // SA FC BC                                           EnergyTotal ExportActiv ImportActiv Volta Curre APowe PFact Frequ Crc--  = DDS238-2 WIFI     
  // SA FC BC EnergyTotal             ExportActiv ImportActiv                         Volta Curre APowe RPowe PFact Frequ Crc--  = DDS238-2 ZN/S version 1 (#6384)
  // SA FC BC EnergyTotal                                     ExportActiv ImportActiv Volta Curre APowe RPowe PFact Frequ Crc--  = DDS238-2 ZN/S version 2 (#6531)
```

So from the above the code to get the data out of the wifi version would be:
```
Energy.voltage[0] = (float)((buffer[29] << 8) + buffer[30]) / 10.0;
Energy.current[0] = (float)((buffer[31] << 8) + buffer[32]) / 100.0;
Energy.active_power[0] = (float)((buffer[33] << 8) + buffer[34]);
Energy.power_factor[0] = (float)((buffer[35] << 8) + buffer[36]) / 1000.0;                                         
Energy.frequency[0] = (float)((buffer[37] << 8) + buffer[38]) / 100.0;                
Energy.export_active = (float)((buffer[21] << 24) + (buffer[22] << 16) + (buffer[23] << 8) + buffer[24]) / 100.0; 
float import_active = (float)((buffer[25] << 24) + (buffer[26] << 16) + (buffer[27] << 8) + buffer[28]) / 100.0;
```

How do I go about adding support, or getting support added for the DDS-238-2-wifi to get reverse energy data out. ?
What who I need to do ?

thanks for any help

**Comment by barbudor** [comment]
Tuya is all about fnId and dpId
It is not because the Modbus version send those information that it would be true for the Wifi version. And even if it is the case, the way informations are reported over Modbus would probably won't give a clue on how they are reported in Tuya protocol.

If you read [TuyaMCU fnId list](https://tasmota.github.io/docs/TuyaMCU/#fnid) you may have noticed that Frequency and Reverse power are not part of the standard fnId codes. Which doesn't mean they are not transmitted but the device developper are generaly free to implement the Tuya protocol in a "flexible" manner.

You likely need to enable `Weblog 3` or even `WebLog 4` to get more detailled alogs about the messages reveived from the TuyaMCU and analyse them to identify is anything that could be the value of interrest are transmitted.
Then some enhancement would need to be added to Tasmota TuyaMCU module itself as today neither of those values are considered by the TuyaMUC driver.

Please share your progress

**Comment by davoau** [comment]
Thanks barbudo for the info.
I have already tried weblog 4, and no info in tuyareceive messages relate to reverse energy and frequency.

```
08:35:01 - tele/tasmota_09C676/RESULT {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
08:35:08 - tele/tasmota_09C676/RESULT {"TuyaReceived":{"Data":"55AA03070008120200040000000029","Cmnd":7,"CmndData":"1202000400000000","DpType2Id18":0,"18":{"DpId":18,"DpIdType":2,"DpIdData":"00000000"}}}
08:35:09 - tele/tasmota_09C676/RESULT {"TuyaReceived":{"Data":"55AA0307000813020004000000002A","Cmnd":7,"CmndData":"1302000400000000","DpType2Id19":0,"19":{"DpId":19,"DpIdType":2,"DpIdData":"00000000"}}}
08:35:10 - tele/tasmota_09C676/RESULT {"TuyaReceived":{"Data":"55AA0307000814020004000009A5D9","Cmnd":7,"CmndData":"14020004000009A5","DpType2Id20":2469,"20":{"DpId":20,"DpIdType":2,"DpIdData":"000009A5"}}}
08:35:11 - tele/tasmota_09C676/RESULT {"TuyaReceived":{"Data":"55AA03070008650200040000000A86","Cmnd":7,"CmndData":"650200040000000A","DpType2Id101":10,"101":{"DpId":101,"DpIdType":2,"DpIdData":"0000000A"}}}
```

With the backlog command used to get the info:
```
Backlog TuyaMCU 33,20; TuyaMCU 32,18; TuyaMCU 31,19; TuyaMCU 37,101;
```
And using tuyaMCU helper bookmarklet:

![Capture2](https://user-images.githubusercontent.com/81886517/169634046-eab61114-fc4d-4c3f-9055-3b2c69bbb0c7.JPG)

The 9 dpID I have been told is a timer.

Is it worth putting in a request to the tuyaMCU driver developer to consider support for reverse energy and frequency, and if so where do I do this ?

**Reply by barbudor** [reply]
You just demonstrated that the TuyaMCU is not sending the information you are praying for to the ESP
So there's nothing Tasmota can do.
Tasmota can only process what the TuyaMCU sends. 

You will have to switch to the MODBUS version to get what you are looking for

**Reply by gomezguillermo** [reply]
I have same device, and I'm looking forward to archive the reverse energy

Checking device logs I'm able to see 9 dpID tuya value 0, right now the meter display is showing -000000, I need to apply some reverse current to see if the value change and if I receive that value in tuya logs. Let you know...

```
09:44:03.552 {"TuyaReceived":{"Data":"55AA03070008090200040000000020","Cmnd":7,"CmndData":"0902000400000000","DpType2Id9":0,"9":{"DpId":9,"DpIdType":2,"DpIdData":"00000000"}}}
09:44:03.570 {"TuyaReceived":{"Data":"55AA0307000865020004000AABCAFB","Cmnd":7,"CmndData":"65020004000AABCA","DpType2Id101":699338,"101":{"DpId":101,"DpIdType":2,"DpIdData":"000AABCA"}}}
09:44:03.589 {"TuyaReceived":{"Data":"55AA0307000812020004000005B6E4","Cmnd":7,"CmndData":"12020004000005B6","DpType2Id18":1462,"18":{"DpId":18,"DpIdType":2,"DpIdData":"000005B6"}}}
09:44:03.608 {"TuyaReceived":{"Data":"55AA03070008130200040000071E4F","Cmnd":7,"CmndData":"130200040000071E","DpType2Id19":1822,"19":{"DpId":19,"DpIdType":2,"DpIdData":"0000071E"}}}
09:44:03.626 {"TuyaReceived":{"Data":"55AA03070008140200040000091A4E","Cmnd":7,"CmndData":"140200040000091A","DpType2Id20":2330,"20":{"DpId":20,"DpIdType":2,"DpIdData":"0000091A"}}}
```
